packages/models/src/store/store.ts

```diff
@@ -10,57 +12,129 @@ import type {
   GetStorageOption,
+  UpdateStorageValue,
+  GetOnChainStorage,
 } from './interface'
 import type { JSONStorageOffChain } from './json-storage'
 import type { GetState } from './chain-storage'
 import { ChainStorage } from './chain-storage'
 import { Actor } from '../actor'
 import { JSONStorage } from './json-storage'
-import { NonExistentException, NonStorageInstanceException } from '../exceptions'
+import {
+  NonExistentCellException,
+  NonExistentException,
+  NonStorageInstanceException,
+  NoSchemaException,
+  UnmatchLengthException,
+} from '../exceptions'
 
 type GetKeyType<T, K extends keyof T> = T & { _add: never } extends { [P in K]: T[P] } ? T[K] : never
 
+export function getUint8ArrayfromHex(value: string, offset: number, length: number) {
+  return bytes.bytify(`0x${value.slice(2 + offset, length ? 2 + offset + length : undefined)}`)
+}
+
 export class Store<
   StorageT extends ChainStorage,
   StructSchema extends StorageSchema<GetState<StorageT>>,
   Option = never,
-> extends Actor<GetStorageStruct<StructSchema>, MessagePayload<StoreMessage<GetStorageStruct<StructSchema>>>> {
+> extends Actor<GetStorageStruct<StructSchema>, MessagePayload<StoreMessage>> {
   protected states: Record<OutPointString, GetStorageStruct<StructSchema>>
+  protected chainData: Record<OutPointString, UpdateStorageValue>
 
   protected options?: Option
 
-  protected schemaOption?: GetStorageOption<StructSchema>
+  protected schemaOption: GetStorageOption<StructSchema>
 
   constructor(
-    params?: GetStorageOption<StructSchema> extends never
-      ? {
-          states?: Record<OutPointString, GetStorageStruct<StructSchema>>
-          options?: Option
-        }
-      : {
-          states?: Record<OutPointString, GetStorageStruct<StructSchema>>
-          options?: Option
-          schemaOption: GetStorageOption<StructSchema>
-        },
+    schemaOption: GetStorageOption<StructSchema>,
+    params?: {
+      states?: Record<OutPointString, GetStorageStruct<StructSchema>>
+      chainData?: Record<OutPointString, UpdateStorageValue>
+      options?: Option
+    },
   ) {
     super()
+    this.schemaOption = schemaOption
     this.states = params?.states || {}
+    this.chainData = params?.chainData || {}
     this.options = params?.options
-    if (params && 'schemaOption' in params) {
-      this.schemaOption = params?.schemaOption
+  }
+
+  private getOffsetAndLength(option: unknown): [number, number] {
+    if (typeof option !== 'object' || option === null) return [0, 0]
+    if (
+      'offset' in option &&
+      typeof option.offset === 'number' &&
+      'length' in option &&
+      typeof option.length === 'number'
+    ) {
+      return [option.offset || 0, option.length || 0]
+    }
+    if ('offset' in option && typeof option.offset === 'number') {
+      return [option.offset || 0, 0]
+    }
+    if ('length' in option && typeof option.length === 'number') {
+      return [0, option.length || 0]
+    }
+    return [0, 0]
```

Comment:
It's prone to miss the branch. I would suggest getting `offset` and `length` first, e.g.

```
const offset = 'offset' in option && typeof option.offset === 'number' ? option.offset : 0
const length = 'length' in option && typeof option.length === 'number' ? option.length : 0
return [offset, length]
```